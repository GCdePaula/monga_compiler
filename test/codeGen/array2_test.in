
arr2 : [[int]];

main () : int {
  i : int;
  k : int;

  arr2 = new [int] [20];
  i = 0;
  while i < 20 {
    arr2[i] = new int [10];
    k = 0;
    while k < 10 {
      arr2[i][k] = i * 10 + k;
      k = k + 1;
    }
    i = i + 1;
  }

  @ "Printing array2";

  i = 0;
  while i < 20 {
    k = 0;
    while k < 10 {
      @ arr2[i][k];
      k = k + 1;
    }
    i = i + 1;
  }

  return 0;
}
