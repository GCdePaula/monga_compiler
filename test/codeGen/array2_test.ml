open Base

let main () =
  let fname = "array2_test.in" in
  Gen.test fname

let _ = Backtrace.Exn.with_recording true ~f:main
