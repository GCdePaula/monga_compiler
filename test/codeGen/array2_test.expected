FILE array2_test.monga LLVM CODE
; ModuleID = 'program'
source_filename = "program"

@0 = global i64** null
@1 = private unnamed_addr constant [16 x i8] c"Printing array2\00", align 1
@2 = private unnamed_addr constant [6 x i8] c"\22%s\22\0A\00", align 1
@3 = private unnamed_addr constant [5 x i8] c"%ld\0A\00", align 1

declare i32 @printf(i8*, ...)

declare i8* @malloc(i64)

define i64 @main() {
entry:
  %0 = alloca i64
  store i64 0, i64* %0
  %1 = alloca i64
  store i64 0, i64* %1
  %2 = call i8* @malloc(i64 mul (i64 ptrtoint (i1** getelementptr (i1*, i1** null, i32 1) to i64), i64 20))
  %3 = bitcast i8* %2 to i64**
  store i64** %3, i64*** @0
  store i64 0, i64* %0
  br label %4

4:                                                ; preds = %29, %entry
  %5 = load i64, i64* %0
  %6 = icmp slt i64 %5, 20
  br i1 %6, label %7, label %32

7:                                                ; preds = %4
  %8 = call i8* @malloc(i64 mul (i64 ptrtoint (i64* getelementptr (i64, i64* null, i32 1) to i64), i64 10))
  %9 = bitcast i8* %8 to i64*
  %10 = load i64**, i64*** @0
  %11 = load i64, i64* %0
  %12 = getelementptr i64*, i64** %10, i64 %11
  store i64* %9, i64** %12
  store i64 0, i64* %1
  br label %13

13:                                               ; preds = %16, %7
  %14 = load i64, i64* %1
  %15 = icmp slt i64 %14, 10
  br i1 %15, label %16, label %29

16:                                               ; preds = %13
  %17 = load i64, i64* %0
  %18 = mul i64 %17, 10
  %19 = load i64, i64* %1
  %20 = add i64 %18, %19
  %21 = load i64**, i64*** @0
  %22 = load i64, i64* %0
  %23 = getelementptr i64*, i64** %21, i64 %22
  %24 = load i64*, i64** %23
  %25 = load i64, i64* %1
  %26 = getelementptr i64, i64* %24, i64 %25
  store i64 %20, i64* %26
  %27 = load i64, i64* %1
  %28 = add i64 %27, 1
  store i64 %28, i64* %1
  br label %13

29:                                               ; preds = %13
  %30 = load i64, i64* %0
  %31 = add i64 %30, 1
  store i64 %31, i64* %0
  br label %4

32:                                               ; preds = %4
  %33 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([6 x i8], [6 x i8]* @2, i32 0, i32 0), i8* getelementptr inbounds ([16 x i8], [16 x i8]* @1, i32 0, i32 0))
  store i64 0, i64* %0
  br label %34

34:                                               ; preds = %52, %32
  %35 = load i64, i64* %0
  %36 = icmp slt i64 %35, 20
  br i1 %36, label %37, label %55

37:                                               ; preds = %34
  store i64 0, i64* %1
  br label %38

38:                                               ; preds = %41, %37
  %39 = load i64, i64* %1
  %40 = icmp slt i64 %39, 10
  br i1 %40, label %41, label %52

41:                                               ; preds = %38
  %42 = load i64**, i64*** @0
  %43 = load i64, i64* %0
  %44 = getelementptr i64*, i64** %42, i64 %43
  %45 = load i64*, i64** %44
  %46 = load i64, i64* %1
  %47 = getelementptr i64, i64* %45, i64 %46
  %48 = load i64, i64* %47
  %49 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([5 x i8], [5 x i8]* @3, i32 0, i32 0), i64 %48)
  %50 = load i64, i64* %1
  %51 = add i64 %50, 1
  store i64 %51, i64* %1
  br label %38

52:                                               ; preds = %38
  %53 = load i64, i64* %0
  %54 = add i64 %53, 1
  store i64 %54, i64* %0
  br label %34

55:                                               ; preds = %34
  ret i64 0
}

RETURN 0

OUTPUT
"Printing array2"
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
